### ISSUE TITLE
Augmented assignment

### ISSUE THREAD

**Issue by Libavius** [j3-fortran/fortran_proposals#113]
allow writing
```
a += b
```
instead of
```
a = a + b
```
and
```
a *= b
```
instead of
```
a = a * b
```

**Comment by certik** [comment]
This is a very common request, so we should have an issue for it and document why such ideas were not implemented yet.

The last time I have seen this discussed is with this proposal: https://j3-fortran.org/doc/year/19/19-111r1.txt. It actually has reason why it was rejected.

**Comment by Libavius** [comment]
> This is a very common request, so we should have an issue for it and document why such ideas were not implemented yet.
> 
> The last time I have seen this discussed is with this proposal: https://j3-fortran.org/doc/year/19/19-111r1.txt. It actually has reason why it was rejected.
> 
> (The `+=` and `*=` would also imply to have `/=`, but that is already taken for `.ne.`. The `19-111r1` paper thus introduces a new operator that could do the same.)

I think, just doing the two 'positive' operators would already help a lot. '-=' and '/=' are much less needed and can be circumvented like this:
The hypothetical code:
``` Fortran
a = 1.0_real64
do i=1, 10
    a -= vector(i)
end do
```
Can be written as:
``` Fortran
a = 1.0_real64
a *= -1
do i=1, 10
    a += vector(i)
end do
a *= -1
```
And analogous for division:
``` Fortran
a = 1.0_real64
a = 1/a; do i=1, 10
    a *= vector(i)
end do; a = 1/a
```
While this seems messy, it prevents the 'space ship operator' from being valid:
```
a -=- 1  ! this gives the same as a += 1
```
The main point however should be, that '-=' and '/=' are much less needed. I would really advise against another operator for '/='. That would be quite counter-intuitive.

PS: since the referred paper was in parts rejected due to a missing use case, I will ad one to the original post =)

**Comment by certik** [comment]
Good point: the proposal would be to only add `+=` and `*=`. I would say the `+=` is probably the most common in practice. I grepped one of my codes for `i = i + 1` and I use just this particular string 7 times. I use `j = j + 1` 8 times, `k = k + 1` 3 times, etc., and that's just a small subset of all usages for `+=`. So having `+=` would simplify dozens of places in my own code for sure.

**Comment by klausler** [comment]
The important aspect of BCPL's assignment operator(s) is that the left-hand side of the assignment is evaluated only once; e.g., if Fortran were to acquire such a feature, in `A(F()) += X`, there would only be one call made to `F`.

**Comment by Libavius** [comment]
> The important aspect of BCPL's assignment operator(s) is that the left-hand side of the assignment is evaluated only once; e.g., if Fortran were to acquire such a feature, in `A(F()) += X`, there would only be one call made to `F`.

I referenced your point in the original post.

**Comment by klausler** [comment]
Additionally, this feature should specifically state whether the augmented assignment is atomic or reducing or disallowed in some circumstances (shared variable in `DO CONCURRENT`, &/or coindexed references).

**Comment by Libavius** [comment]
I'd suggest also adding `**=` to have the 'positive' augmented assignment operators complete.

**Comment by jacobwilliams** [comment]
For once I agree with the committee for not adding something. 😃 I think these variants are best left to C-based languages. I don't think they are very "Fortranic". Aesthetically I've always thought they were weird, and I don't think we really gain much by having them (except new confusion about `/=`).

**Comment by FortranFan** [comment]
@Libavius wrote:

> ..
> The hypothetical code:
> 
> ```fortran
> a = 1.0_real64
> do i=1, 10
>     a -= vector(i)
> end do
> ```
> 
> Can be written as:
> 
> ```fortran
> a = 1.0_real64
> a *= -1
> do i=1, 10
>     a += vector(i)
> end do
> a *= -1
> ```


The above "hypothetical code" can be written using current Fortran standard better as:

```Fortran
a = 1.0_real64 - sum(vector)
```

**Comment by certik** [comment]
In fact you can write it as
```fortran
a = 1 - sum(vector)
```

But I think @Libavius's point was to show how it can be used. A typical loop where I use `i = i + 1` is:
```fortran
i = 1
j = 1
do j = 1, size(nl)
    n = nl(j)
    zeta = zl(j)
    l = ll(j)
    do m = -l, l
        nlist(i) = n
        zetalist(i) = zeta
        llist(i) = l
        mlist(i) = m
        i = i + 1
    end do
end do
```
Where it is kept as a running index. 

I am leaning towards @jacobwilliams's comment above (https://github.com/j3-fortran/fortran_proposals/issues/113#issuecomment-566687141) that `i = i + 1` looks great and even though `i += 1` is shorter, it is a new syntax, and so it might not be worth it in this case.

**Comment by Libavius** [comment]
@FortranFan I know, thanks anyway for the hint.
@certik that was my point indeed. I also agree, that `i = i + 1` looks good, however this is not always the case.

**A more elaborate use case:**
``` Fortran
type Molecule_t
    type(Atom_t), allocatable :: atoms(:)
contains
    procedure, pass :: Get_electrons_number => Molecule_t_get_electrons_number
end type Molecule_t

contains

function Molecule_t_get_electrons_number(this) result(electrons_number)
    class(Molecule_t), intent(in) :: this
    integer :: electrons_number

    integer :: i

    electrons_number = 0

    if (ALLOCATED(this%atoms) then
        do i=1, SIZE(this%atoms)
            electrons_number = electrons_number + this%atoms(i)%get_electrons_number()
        end do
    end if
end subroutine Molecule_t_get_electrons_number
```
The loop would–in my opinion–be easier readable like this:
``` Fortran
do i=1, SIZE(this%atoms)
    electrons_number += this%atoms(i)%get_electrons_number()
end do
```

@FortranFan I could write this with the `SUM` intrinsic (but this would be slower, since an array would need to be allocated):
``` Fortran
electrons_number = SUM([( this%atoms(i)%get_electrons_number(), i=1, SIZE(this%atoms) )])
```

This is in addition to the possible reduction of function calls that @klausler mentioned.

@jacobwilliams jup, the confusion with `/=` ist quite painful and probably the strongest counter-argument...

**Comment by certik** [comment]
@Libavius for your latest use case, I would recommend to change `get_electrons_number()` to not be a function, but just a member variable. In fact I would recommend to change `Molecule_t` to:
```fortran
type Molecule_t
    integer, allocatable :: electrons_number(:)
    ...
contains
    procedure, pass :: Get_electrons_number => Molecule_t_get_electrons_number
end type Molecule_t
```
And then change the loop to just:
```fortran
electrons_number = sum(this%electrons_number)
```

**Comment by Libavius** [comment]
@certik 
But then, if I change my atoms vector I always have to be careful and update electrons_number as well.

The idea in this case would be, that a molecule has atoms but an atom can also exist without a molecule. To avoid code duplication I don't want the molecule to have a memeber 'electrons_number' or implement atom-related functions again but rather iterate over its atoms.

**Comment by certik** [comment]
The way I do this in my electronic structure codes is that I do not have an Atom type at all. So if your molecule has only one atom, then the length of all the arrays will be just 1. There is no code duplication and in fact things run much faster.

**Comment by septcolor** [comment]
@Libavius wrote:
> But then, if I change my atoms array I always have to be careful and update electrons_number as well.

Apart from the efficiency issue, I think the above case is one motivation for some other languages (like C#, Swift, and also Python (?)) to provide "computed properties" (~ a variant of setter/getter). I also do a similar thing when a type component depends dynamically on other data and so varies "in time". Otherwise, I think we may need to keep track of the update status of such components, e.g. by some flags (in my case, I use both approaches). We can also make a local variable by calling computed properties first (e.g., before entering a heavy loop) and use the local variable in a heavy loop (to avoid the efficiency issue). On-the-fly computation is also sometimes useful (IMO) to make the "responsibility" of underlying data clear (so that only one basic component is responsible for other dependent properties).

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/properties
https://docs.swift.org/swift-book/LanguageGuide/Properties.html

**Comment by cmacmackin** [comment]
@septcolor 
I've also suggested something like these be included in Fortran (#30 ). I feel they are useful because they allow you to expose derived type components, safe in the knowledge that if you need to refactor the data storage layout you can always add a getter in future. They are thus extermely useful for encapsulation, which @certik's approach breaks.

**Comment by septcolor** [comment]
Apart from that, my understanding when "+=" etc become useful is the situation where the name of an entity on the left-hand side is very long, particularly when it is a component of nested derived types. (I remember there were several repeated discussions on this, but cannot find the threads right now...) For example, something like

```
my_data(idat) % its_sub_obj(iobj) % another_sub_obj(isubobj) = my_data(idat) % its_sub_obj(iobj) % another_sub_obj(isubobj) + 1
```
(In the above, I have used "idat" etc to distinguish indices for different data types.)  I experience this situation rather often when I use OO-style coding. Because it is too tedious to write, I use a preprocessor to convert a nonstandard short-hand notation (in my Fortran-like sources) into the original form above. If the standard provides "+=" and "*=", I would of course use it (rather than relying on a local converter/preprocessor approach).

Also, I feel "-=" and "/=" are not particularly necessary, because they can be expressed by "+=" and "*=". In my codes, the former appear much less frequently (maybe close to none?) than the latter.

(By the way, on comp.lang.fortran, there was a discussion that ASSOCIATE could be used to avoid typing a long entity name twice, e.g.,
```
associate( x => my_data(idat) % its_sub_obj(iobj) % another_sub_obj(isubobj) ) 
x = x + 1
end associate
```
but I feel it is not very readable than the more usual approach in other languages
```
my_data(idat) % its_sub_obj(iobj) % another_sub_obj(isubobj) += 1
```
(I used "more usual" because "+=" etc has been used in a basic tool like bash and I use them pretty often.)

**Comment by certik** [comment]
@cmacmackin indeed, the approach I suggested is flat and exposed. It's not encapsulated. It has the huge advantage that it works well with current Fortran, and typically optimizes really well and runs fast, and it's actually typically simpler in terms of lines of code (as demonstrated above). The disadvantage is that if you change the underlying representation, you need to rework your whole code. So with my approach, one has to design things well, but one can use the current Fortran language. For the encapsulated approach, Fortran does not have the best tools currently.

**Comment by aradi** [comment]
As for the syntax, what about Espen Myklebusts proposal in #80, using the `@` operator representing the LHS of an assignment? So @septcolor s example could be writen as
```
my_data(idat) % its_sub_obj(iobj) % another_sub_obj(isubobj) = @ + 1
```
It would not cause any conflicts with current operators, so it could be also used for division (I would find it really strange, if `*=` were introduced but in place division won't work...). And it had the benefit, that the `@` operator (representing the LHS) could appear even more than once on the RHS, as in
```
array(i,j)%comp1(k,l)%value1 = abs(@ - sqrt(@))
```
(example by Espen Myklebust).

**Comment by certik** [comment]
@aradi is your idea different to the proposal I linked at https://github.com/j3-fortran/fortran_proposals/issues/113#issuecomment-566223095, which was rejected?

**Comment by aradi** [comment]
@certik Yes, Espen Myklebusts proposal has in my opinion a considerably lower complexity, as the one you have linked in. The `@` operator would always represent the entire LHS, while in your proposal it can also represent a part of it (as in `a ( s @ (3*i+1) ) = b(s)`). It does not involve naming, so no conflicts with eventual implicit statements must be feared and there is no need to introduce a new (statement level) scope.

**Comment by certik** [comment]
@aradi thanks for the clarification. (For the record the `19-111r1` proposal is not my proposal, I just linked to it.)

**Comment by Libavius** [comment]
> As for the syntax, what about Espen Myklebusts proposal in #80, using the `@` operator representing the LHS of an assignment? So @septcolor s example could be writen as
> 
> ```
> my_data(idat) % its_sub_obj(iobj) % another_sub_obj(isubobj) = @ + 1
> ```
> 
> It would not cause any conflicts with current operators, so it could be also used for division (I would find it really strange, if `*=` were introduced but in place division won't work...). And it had the benefit, that the `@` operator (representing the LHS) could appear even more than once on the RHS, as in
> 
> ```
> array(i,j)%comp1(k,l)%value1 = abs(@ - sqrt(@))
> ```
> 
> (example by Espen Myklebust).

I think this'd be superb elegant =)

**Comment by klausler** [comment]
I like the idea of using a symbol in the right-hand side to denote the current value of the left-hand side, but `@` is accepted as a legal character in identifiers as an extension in many compilers (as is `$` as well).  `#` would conflict with preprocessing directives if it were used and it appeared at the beginning of a continuation line.  '*' would be very Fortran-like but would lead to ambiguity (`X=Y*****Z`), as would `/`.  So that leaves us `%` and `?`, I think.

**Comment by everythingfunctional** [comment]
I think the reasons outlined why augmented assignment probably won't work out are true. But I would be in favor of the "pronoun" proposal that @certik mentioned. There are a variety of languages starting to support this kind of feature (most commonly in languages that have a pattern matching feature). One might think of this as basically syntactic sugar for an associate block with only a single line in it.

For example:
```
v @ x(3*i+1) = v * 42 + b(i)
```
would be directly equivalent to:
```
associate(v => x(3*i+1))
    v = v * 42 + b(i)
end associate
```

So this could basically be treated as syntactic sugar, not really a new functionality.

**Comment by klausler** [comment]
> I think the reasons outlined why augmented assignment probably won't work out are true. But I would be in favor of the "pronoun" proposal that @certik mentioned. There are a variety of languages starting to support this kind of feature (most commonly in languages that have a pattern matching feature). One might think of this as basically syntactic sugar for an associate block with only a single line in it.

`ASSOCIATE()` could be extended slightly to accommodate a single statement as well as a block.

**Comment by septcolor** [comment]
@klausler wrote:
> (...snip...) So that leaves us `%` and `?`, I think.

If Fortran could introduce new symbols (?), I wonder if the tilde (~) may be another possibility (e.g. "long_lhs_expr = ~ + 1"). I also imagined that a single underscore (_) may be another choice, but it might look like "discard" for people familiar with Python etc (like discarding some part in a tuple or a unused loop index).

If a new symbol for augmented assignments is acceptable, I guess something like "long_lhs_expr (+)= 1" might also another possibility.

**Comment by opeil** [comment]
Perhaps, it is too late to discuss this but somehow I have not seen anyone mentioning another important aspect of the compound/augmented assignment: It allows to perform more efficient in-place operations, such as ```big_array(:, :, :) += big_array2(:, :, :)```, without the risk of creating temporaries. This makes it more than just a syntactic sugar.

Another similar example is that if we have a type like this
```
type :: BigObject
   real(dp) :: values(10000, 10000)
end type

type(BigObject) :: a, b
```
and there is a frequent need to do operations like this -- ```a % values = a % values + b % values```, then a naive overloading of operator (+) would result in an extremely inefficient code. It would be great if one has a possibility to overload something like (+=) [with obvious call semantics] to implement heavy in-place operations.

**Comment by septcolor** [comment]
This is another similar request that I have come across on the StackOverflow site:

Increment value in Fortran?
https://stackoverflow.com/questions/64791017/increment-value-in-fortran?noredirect=1&lq=1

**Comment by klausler** [comment]
Some further thought about the use of a special symbol (say `..` here) to represent the entire previous value of the left-hand side of an assignment-stmt is leading me to think that it would be a worthwhile feature to consider.

* It handles the original important use case of `+=`, as `x = .. + y`.
* It works for any operator, including division, concatenation, logicals, and user-defined dyadics, without ambiguity.
* It allows reversed operations (subtract from, divide into, prepend).
* It works as well for operations that are not operators, including subscripts (`ints(:) = permute(..)`), function reference actual arguments, and even designators.
* It allows for multiple occurrences of the left-hand side variable on the right-hand side.
* It could/should be defined to work for left-hand side variables that `ASSOCIATE` *can't* handle (viz., vector-valued subscripts & coindexed variables.
* It looks (to me) less likely than `+=` to imply atomicity to the naive reader.
* It would allow for the use of parentheses to control rounding and perhaps prevent the use of FMA (`x = (.. * y) + z`).

**Comment by certik** [comment]
Another discussion of this proposal: https://fortran-lang.discourse.group/t/updating-assignment-operators/2129

**Comment by kbauer** [comment]
In the renewed discussion mentioned in  https://github.com/j3-fortran/fortran_proposals/issues/113#issuecomment-905758722, there appears the idea that

    A /= B

might actually work, since assignment isn't allowed in expressions, so there should be little risk of mixing up 

    IF( A /= B ) THEN ! or
    L = A /= B

with

    A /= B

as assignment for both compilers and human code readers. While not ideal, I'd take such a solution gladly over lines such as 

    forceFromExternalSource = scaleFactor * forceFromExternalSource
